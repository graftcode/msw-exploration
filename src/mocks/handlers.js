import { rest } from "msw";

export const handlers = [
  // Handles a POST /login request
  rest.get("https://www.omdbapi.com/", (req, res, ctx) => {
    const page = req.url.searchParams.get("page");
    const searchTerm = req.url.searchParams.get("s");

    console.log({
      page,
      searchTerm,
    });

    // Persist user's authentication in the session
    return res(
      // Respond with a 200 status code
      ctx.status(200),
      ctx.json({
        Search: [
          {
            Title: "The ABC Murders",
            Year: "2018",
            imdbID: page + "tt8463714",
            Type: "series",
            Poster:
              "https://m.media-amazon.com/images/M/MV5BYmJmNzU0NTgtOGM5Ni00MWJiLTgwZmMtNDIyNjYzZDAzNzI5XkEyXkFqcGdeQXVyMjExMjk0ODk@._V1_SX300.jpg",
          },
          {
            Title: "ABC Afterschool Specials",
            Year: "1972–1997",
            imdbID: page + "tt0202179",
            Type: "series",
            Poster:
              "https://m.media-amazon.com/images/M/MV5BOTY2MmRiZjQtM2MyMC00MDViLThhMjUtYzZiN2JkYzA4NDE5XkEyXkFqcGdeQXVyMTY2MzYyNzA@._V1_SX300.jpg",
          },
          {
            Title: "ABC Africa",
            Year: "2001",
            imdbID: page + "tt0281534",
            Type: "movie",
            Poster:
              "https://m.media-amazon.com/images/M/MV5BMTU5ODc1ODc2Ml5BMl5BanBnXkFtZTcwMzEyOTgyMQ@@._V1_SX300.jpg",
          },
          {
            Title: "ABC TGIF",
            Year: "1989–2018",
            imdbID: page + "tt0312324",
            Type: "series",
            Poster:
              "https://m.media-amazon.com/images/M/MV5BMDY2M2MyZTEtYmZlZi00NTUyLWE4MWItNDg3YmIzY2VjZWYyXkEyXkFqcGdeQXVyMTEwODg2MDY@._V1_SX300.jpg",
          },
          {
            Title: "ABC World News Tonight with David Muir",
            Year: "1953–",
            imdbID: page + "tt0184090",
            Type: "series",
            Poster:
              "https://m.media-amazon.com/images/M/MV5BYWZjYWYyMjctMjIyMS00YzM5LWExYzctZmI4OWFmZDQyNGQ1XkEyXkFqcGdeQXVyNzU5OTQ1MDM@._V1_SX300.jpg",
          },
          {
            Title: "ABC News Nightline",
            Year: "1980–",
            imdbID: page + "tt0154053",
            Type: "series",
            Poster:
              "https://m.media-amazon.com/images/M/MV5BYTNiODgzYWItYmYyNi00NDk1LWEzOGUtNjUyZjg3NmU5YTJhXkEyXkFqcGdeQXVyODg3NDc1OTE@._V1_SX300.jpg",
          },
          {
            Title: "ABC Weekend Specials",
            Year: "1977–1995",
            imdbID: page + "tt0075471",
            Type: "series",
            Poster:
              "https://m.media-amazon.com/images/M/MV5BNjllM2ViM2MtNTJkYi00YjZhLTkyMjEtMzBkZWU1OGQzYTA0XkEyXkFqcGdeQXVyMTg2NjYzOA@@._V1_SX300.jpg",
          },
          {
            Title: "Garotas do ABC",
            Year: "2003",
            imdbID: page + "tt0327451",
            Type: "movie",
            Poster:
              "https://m.media-amazon.com/images/M/MV5BZjdkMTkyMTMtMmU1MC00MDYyLWEwMWUtMDdiYzYwNmI0ZjZiXkEyXkFqcGdeQXVyMTU2ODc4ODQ@._V1_SX300.jpg",
          },
          {
            Title: "ABC da Greve",
            Year: "1990",
            imdbID: page + "tt0317051",
            Type: "movie",
            Poster:
              "https://m.media-amazon.com/images/M/MV5BNDMxN2FjMjAtNTQ4YS00MWE0LWIzMjktZjc0YTUxOTc5YWFkXkEyXkFqcGdeQXVyOTU3ODk4MQ@@._V1_SX300.jpg",
          },
          {
            Title: "ABC Late Night",
            Year: "1973–1976",
            imdbID: page + "tt0176356",
            Type: "series",
            Poster:
              "https://m.media-amazon.com/images/M/MV5BNzQxOTAyYTMtODczNy00MWUxLTlkY2YtY2VkOGVmNmM3YmIxXkEyXkFqcGdeQXVyMjcyMDU4NA@@._V1_SX300.jpg",
          },
        ],
        totalResults: "30",
        Response: "True",
      })
    );
  }),
  rest.get("https://m.media-amazon.com/", (req, res, ctx) => {
    // Persist user's authentication in the session
    return res(
      // Respond with a 200 status code
      ctx.status(200)
    );
  }),
];
