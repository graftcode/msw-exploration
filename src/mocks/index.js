export { movieInformationMock } from "./movieInformationMock";
export { searchMoviesMock } from "./searchMoviesMock";
