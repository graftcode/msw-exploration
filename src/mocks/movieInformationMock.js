export const movieInformationMock = () => ({
  Title: "The ABC Murders",
  Year: "2018",
  Rated: "TV-14",
  Released: "01 Feb 2019",
  Runtime: "2 min",
  Genre: "Crime, Drama, Mystery",
  Director: "N/A",
  Writer: "N/A",
  Actors: "John Malkovich, Eamon Farren, Michael Shaeffer",
  Plot: 'In 1933, retired detective Hercule Poirot is targeted by a taunting killer who sends letters signed "ABC", which Poirot must decode in order to discover the identity of the murderer.',
  Language: "English",
  Country: "United Kingdom",
  Awards: "3 nominations",
  Poster:
    "https://m.media-amazon.com/images/M/MV5BYmJmNzU0NTgtOGM5Ni00MWJiLTgwZmMtNDIyNjYzZDAzNzI5XkEyXkFqcGdeQXVyMjExMjk0ODk@._V1_SX300.jpg",
  Ratings: [
    {
      Source: "Internet Movie Database",
      Value: "6.6/10",
    },
  ],
  Metascore: "N/A",
  imdbRating: "6.6",
  imdbVotes: "10,579",
  imdbID: "tt8463714",
  Type: "series",
  totalSeasons: "1",
  Response: "True",
});
