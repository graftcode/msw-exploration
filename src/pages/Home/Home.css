.search-movies-form {
  margin-bottom: 32px;
  font-size: 16px;
  padding: 10px;
  border: solid 1px;
  display: grid;
  grid-template-columns: 50px 1fr;
  column-gap: 16px;
  overflow: hidden;
}

.search-movies-form button {
  border: none;
  background: none;
  font-size: 32px;
  cursor: pointer;
}

.search-movies-form input {
  border: none;
  outline: none;
}

.search-movies-form button:active {
  transform: scale(0.8);
}

.search-movies-result {
  display: grid;
  column-gap: 32px;
  row-gap: 32px;
  margin-bottom: 32px;
}

@media (min-width: 425px) {
  .search-movies-result {
    grid-template-columns: 1fr 1fr;
  }
}

@media (min-width: 767px) {
  .search-movies-result {
    grid-template-columns: 1fr 1fr 1fr;
  }
}

.movie-item {
  padding: 16px;
  border: solid 2px black;
  border-radius: 10px;
  cursor: pointer;
  text-align: center;
  font-weight: 700;
  font-size: 24px;
}

.movie-item img {
  height: 300px;
  width: 100%;
  margin-bottom: 16px;
}

/* svg shown when using mocks */
.movie-icon {
  width: 100%;
}

.movie-item:hover {
  background: #60a5fa;
  border: solid 2px #60a5fa;
}

.load-more-button {
  background: none;
  border-radius: 5px;
  cursor: pointer;
  display: block;
  font-size: 24px;
  font-weight: 700;
  margin: auto;
  padding: 10px 20px;
}

.load-more-button:hover {
  color: #60a5fa;
  border: solid 2px #60a5fa;
}
