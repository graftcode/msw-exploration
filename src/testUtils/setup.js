import { render } from "@testing-library/react";
import { MemoryRouter, Route } from "react-router-dom";

export const setup = (Component) => {
  return render(
    <MemoryRouter>
      <Component />
    </MemoryRouter>
  );
};
