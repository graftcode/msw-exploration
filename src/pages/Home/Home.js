import React, { useState } from "react";
import { Link } from "react-router-dom";

import { getMovies } from "../../axios";
import NoImageAvailable from "../../assets/NoImageAvailable.png";

import "./Home.css";

const Home = () => {
  const [searchTerm, setSearchTerm] = useState("");
  const [moviesData, setMoviesData] = useState([]);
  const [pageNumber, setPageNumber] = useState(1);
  const [numberOfPages, setNumberOfPages] = useState(1);
  const [errorMessage, setErrorMessage] = useState("");

  const resetState = () => {
    setMoviesData([]);
    setPageNumber(1);
    setNumberOfPages(1);
    setErrorMessage("");
  };

  const handleSubmit = async (e) => {
    e.preventDefault();

    resetState();

    if (!searchTerm) return;

    const { data, error } = await getMovies(searchTerm, pageNumber);

    if (error || data.Error) {
      setErrorMessage(error?.message || data?.Error);
      setMoviesData([]);
      return;
    }

    setMoviesData(data?.Search);
    setNumberOfPages(Math.ceil(Number(data?.totalResults) / 10));
  };

  const handleLoadMore = async () => {
    const { data, error } = await getMovies(searchTerm, pageNumber + 1);

    if (error) {
      setErrorMessage(error.message);
      return;
    }

    if (errorMessage) {
      setErrorMessage("");
    }

    setPageNumber(pageNumber + 1);
    setMoviesData([...moviesData, ...data?.Search]);
  };

  return (
    <section className="search-movies">
      <form className="search-movies-form" onSubmit={handleSubmit}>
        <button className="search-movies-form-button">🔍</button>
        <input
          type="text"
          placeholder="Search for a movie..."
          onChange={(e) => setSearchTerm(e.target.value)}
        />
      </form>

      <div className="search-movies-result">
        {moviesData.length > 0 &&
          moviesData.map((movie) => (
            <div className="movie-item" key={movie.imdbID}>
              <img
                src={movie.Poster}
                onError={({ currentTarget }) => {
                  currentTarget.onerror = null;
                  currentTarget.src = NoImageAvailable;
                }}
                alt={movie.Title}
              />
              <Link to={`${movie.imdbID}`} className="link">
                {movie.Title}
              </Link>
            </div>
          ))}
      </div>
      {pageNumber !== numberOfPages && (
        <button className="load-more-button" onClick={handleLoadMore}>
          Load More
        </button>
      )}
      {errorMessage && <p className="error-message">{errorMessage}</p>}
    </section>
  );
};

export default Home;
