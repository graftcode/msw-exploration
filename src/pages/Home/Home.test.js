import { fireEvent, screen, waitFor } from "@testing-library/react";
import { setupServer } from "msw/node";
import { rest } from "msw";

import Home from "./Home";

import { searchMoviesMock } from "../../mocks/searchMoviesMock";
import { setup } from "../../testUtils/setup";

const handlers = [
  rest.get("https://www.omdbapi.com", (req, res, ctx) => {
    const page = req.url.searchParams.get("page");
    const searchTerm = req.url.searchParams.get("s");

    if (page && searchTerm) return res(ctx.json(searchMoviesMock(page)));
  }),

  rest.get("https://m.media-amazon.com/*", async (req, res, ctx) => {
    return res(ctx.set("Content-Type", "image/svg+xml"));
  }),
];

const server = setupServer(...handlers);

beforeEach(() => {
  setup(Home);
});

beforeAll(() => {
  server.listen();
});

afterEach(() => server.resetHandlers());

afterAll(() => {
  server.close();
});

describe("<Home />", () => {
  test("renders Home component", () => {
    const testComponent = screen.getByTestId("search-movies");
    expect(testComponent).toBeInTheDocument();
  });

  describe("Search movie form", () => {
    test("renders Input component", () => {
      const inputComponent = screen.getByPlaceholderText(
        "Search for a movie..."
      );
      expect(inputComponent).toBeInTheDocument();
    });

    test("renders searchButton component", () => {
      const searchButton = screen.getByText("🔍");
      expect(searchButton).toBeInTheDocument();
    });

    test("renders a list on movies when form is submitted successfully", async () => {
      const inputComponent = screen.getByPlaceholderText(
        "Search for a movie..."
      );
      const searchButton = screen.getByText("🔍");

      fireEvent.change(inputComponent, { target: { value: "abc" } });
      fireEvent.click(searchButton);

      // Wait for the API request to complete
      // then make assertions
      await waitFor(() => {
        const movieItems = screen.getAllByTestId("movie-item");

        expect(movieItems.length).toBe(10);
        expect(screen.getAllByTestId("search-movies-result")).toMatchSnapshot();
      });
    });
  });
});
