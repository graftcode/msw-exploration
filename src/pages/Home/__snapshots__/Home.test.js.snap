// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> Search movie form renders a list on movies when form is submitted successfully 1`] = `
Array [
  <div
    class="search-movies-result"
    data-testid="search-movies-result"
  >
    <div
      class="movie-item"
      data-testid="movie-item"
    >
      <img
        alt="The ABC Murders"
        src="https://m.media-amazon.com/images/M/MV5BYmJmNzU0NTgtOGM5Ni00MWJiLTgwZmMtNDIyNjYzZDAzNzI5XkEyXkFqcGdeQXVyMjExMjk0ODk@._V1_SX300.jpg"
      />
      <a
        class="link"
        data-testid="link"
        href="/1tt8463714"
      >
        The ABC Murders
      </a>
    </div>
    <div
      class="movie-item"
      data-testid="movie-item"
    >
      <img
        alt="ABC Afterschool Specials"
        src="https://m.media-amazon.com/images/M/MV5BOTY2MmRiZjQtM2MyMC00MDViLThhMjUtYzZiN2JkYzA4NDE5XkEyXkFqcGdeQXVyMTY2MzYyNzA@._V1_SX300.jpg"
      />
      <a
        class="link"
        data-testid="link"
        href="/1tt0202179"
      >
        ABC Afterschool Specials
      </a>
    </div>
    <div
      class="movie-item"
      data-testid="movie-item"
    >
      <img
        alt="ABC Africa"
        src="https://m.media-amazon.com/images/M/MV5BMTU5ODc1ODc2Ml5BMl5BanBnXkFtZTcwMzEyOTgyMQ@@._V1_SX300.jpg"
      />
      <a
        class="link"
        data-testid="link"
        href="/1tt0281534"
      >
        ABC Africa
      </a>
    </div>
    <div
      class="movie-item"
      data-testid="movie-item"
    >
      <img
        alt="ABC TGIF"
        src="https://m.media-amazon.com/images/M/MV5BMDY2M2MyZTEtYmZlZi00NTUyLWE4MWItNDg3YmIzY2VjZWYyXkEyXkFqcGdeQXVyMTEwODg2MDY@._V1_SX300.jpg"
      />
      <a
        class="link"
        data-testid="link"
        href="/1tt0312324"
      >
        ABC TGIF
      </a>
    </div>
    <div
      class="movie-item"
      data-testid="movie-item"
    >
      <img
        alt="ABC World News Tonight with David Muir"
        src="https://m.media-amazon.com/images/M/MV5BYWZjYWYyMjctMjIyMS00YzM5LWExYzctZmI4OWFmZDQyNGQ1XkEyXkFqcGdeQXVyNzU5OTQ1MDM@._V1_SX300.jpg"
      />
      <a
        class="link"
        data-testid="link"
        href="/1tt0184090"
      >
        ABC World News Tonight with David Muir
      </a>
    </div>
    <div
      class="movie-item"
      data-testid="movie-item"
    >
      <img
        alt="ABC News Nightline"
        src="https://m.media-amazon.com/images/M/MV5BYTNiODgzYWItYmYyNi00NDk1LWEzOGUtNjUyZjg3NmU5YTJhXkEyXkFqcGdeQXVyODg3NDc1OTE@._V1_SX300.jpg"
      />
      <a
        class="link"
        data-testid="link"
        href="/1tt0154053"
      >
        ABC News Nightline
      </a>
    </div>
    <div
      class="movie-item"
      data-testid="movie-item"
    >
      <img
        alt="ABC Weekend Specials"
        src="https://m.media-amazon.com/images/M/MV5BNjllM2ViM2MtNTJkYi00YjZhLTkyMjEtMzBkZWU1OGQzYTA0XkEyXkFqcGdeQXVyMTg2NjYzOA@@._V1_SX300.jpg"
      />
      <a
        class="link"
        data-testid="link"
        href="/1tt0075471"
      >
        ABC Weekend Specials
      </a>
    </div>
    <div
      class="movie-item"
      data-testid="movie-item"
    >
      <img
        alt="Garotas do ABC"
        src="https://m.media-amazon.com/images/M/MV5BZjdkMTkyMTMtMmU1MC00MDYyLWEwMWUtMDdiYzYwNmI0ZjZiXkEyXkFqcGdeQXVyMTU2ODc4ODQ@._V1_SX300.jpg"
      />
      <a
        class="link"
        data-testid="link"
        href="/1tt0327451"
      >
        Garotas do ABC
      </a>
    </div>
    <div
      class="movie-item"
      data-testid="movie-item"
    >
      <img
        alt="ABC da Greve"
        src="https://m.media-amazon.com/images/M/MV5BNDMxN2FjMjAtNTQ4YS00MWE0LWIzMjktZjc0YTUxOTc5YWFkXkEyXkFqcGdeQXVyOTU3ODk4MQ@@._V1_SX300.jpg"
      />
      <a
        class="link"
        data-testid="link"
        href="/1tt0317051"
      >
        ABC da Greve
      </a>
    </div>
    <div
      class="movie-item"
      data-testid="movie-item"
    >
      <img
        alt="ABC Late Night"
        src="https://m.media-amazon.com/images/M/MV5BNzQxOTAyYTMtODczNy00MWUxLTlkY2YtY2VkOGVmNmM3YmIxXkEyXkFqcGdeQXVyMjcyMDU4NA@@._V1_SX300.jpg"
      />
      <a
        class="link"
        data-testid="link"
        href="/1tt0176356"
      >
        ABC Late Night
      </a>
    </div>
  </div>,
]
`;
